File: tools/lvtools/compression.py
from array import array
import itertools

from util import read_word

def decompress_data(data, decoded_data_len=None):
    if decoded_data_len is None:
        decoded_data_len = read_word(data, 0) + 1
        data = data[2:]

    scratch = array('B', itertools.repeat(0, 0x1000))
    dst = array('B')

    scratch_i = 0
    src_i = 0

    while True:
        al = data[src_i]
        src_i += 1

        for i in xrange(8):
            if al & 1:
                # Verbatim byte
                scratch[scratch_i] = data[src_i]
                scratch_i = (scratch_i + 1) % len(scratch)
                dst.append(data[src_i])
                if len(dst) == decoded_data_len:
                    return dst
                src_i += 1
            else:
                # Copy previous bytes from scratch area
                copy_src = read_word(data, src_i)
                src_i += 2

                copy_len = (copy_src >> 12) + 3
                copy_src = copy_src & 0xFFF

                for j in xrange(copy_len):
                    scratch[scratch_i] = scratch[copy_src]
                    scratch_i = (scratch_i + 1) % len(scratch)
                    dst.append(scratch[copy_src])
                    if len(dst) == decoded_data_len:
                        return dst
                    copy_src = (copy_src + 1) % len(scratch)
            al = al >> 1

File: tools-ng/src/lvtools/compression.rs
use std::vec;

pub fn decompress_data<T: Reader>(src: &mut T, decoded_data_len: uint) -> ~[u8] {
	let mut scratch = [0u8, ..4*1024];
	let mut dst = vec::with_capacity(decoded_data_len);

	let mut scratch_i = 0;

	loop {
		let mut al = src.read_u8();

		for _ in range(0, 8) {
			if al & 1 != 0 {
				// Verbatim byte
				let symbol = src.read_u8();

				scratch[scratch_i] = symbol;
				scratch_i = (scratch_i + 1) % scratch.len();

				dst.push(symbol);
				if dst.len() == decoded_data_len {
					return dst;
				}
			} else {
				// Copy bytes from scratch area
				let copy_op = src.read_le_u16();

				let copy_len = ((copy_op >> 12) + 3) as uint;
				let mut copy_src = (copy_op & 0xFFF) as uint;

				for _ in range(0, copy_len) {
					let symbol = scratch[copy_src];
					copy_src = (copy_src + 1) % scratch.len();

					scratch[scratch_i] = symbol;
					scratch_i = (scratch_i + 1) % scratch.len();

					dst.push(symbol);
					if dst.len() == decoded_data_len {
						return dst;
					}
				}
			}

			al = al >> 1;
		}
	}
}
